crates/iota-core/src/starfish_adapter.rs

```diff
@@ -0,0 +1,147 @@
+        }
+    }
+
+    async fn get(&self) -> Guard<Option<Arc<TransactionClient>>> {
+        let client = self.client.load();
+        if client.is_some() {
+            return client;
+        }
+
+        // Consensus client is initialized after validators or epoch starts, and cleared
+        // after an epoch ends. But calls to get() can happen during validator
+        // startup or epoch change, before consensus finished initializations.
+        // TODO: maybe listen to updates from consensus manager instead of polling.
+        let mut count = 0;
+        let start = Instant::now();
+        const RETRY_INTERVAL: Duration = Duration::from_millis(100);
+        loop {
+            let client = self.client.load();
+            if client.is_some() {
+                return client;
+            } else {
+                sleep(RETRY_INTERVAL).await;
+                count += 1;
+                if count % 100 == 0 {
+                    warn!(
+                        "Waiting for consensus to initialize after {:?}",
+                        Instant::now() - start
+                    );
+                }
+            }
+        }
+    }
+
+    pub fn set(&self, client: Arc<TransactionClient>) {
+        self.client.store(Some(client));
+    }
+
+    pub fn clear(&self) {
+        self.client.store(None);
+    }
+}
+
+#[async_trait::async_trait]
+impl ConsensusClient for LazyStarfishClient {
+    async fn submit(
+        &self,
+        transactions: &[ConsensusTransaction],
+        _epoch_store: &Arc<AuthorityPerEpochStore>,
+    ) -> IotaResult<BlockStatusReceiver> {
+        // TODO(starfish): confirm comment is still true
+        // The retrieved TransactionClient can be from the past epoch. Submit would fail
+        // after Starfish shuts down, so there should be no correctness issue.
+        let client = self.get().await;
+        let transactions_bytes = transactions
+            .iter()
+            .map(|t| bcs::to_bytes(t).expect("Serializing consensus transaction cannot fail"))
+            .collect::<Vec<_>>();
+
+        let (block_ref, status_waiter) = client
+            .as_ref()
+            .expect("Client should always be returned")
+            .submit(transactions_bytes)
+            .await
+            .tap_err(|err| {
+                // Will be logged by caller as well.
+                let msg = format!("Transaction submission failed with: {:?}", err);
+                match err {
+                    ClientError::ConsensusShuttingDown(_) => {
+                        info!("{}", msg);
+                    }
+                    ClientError::OversizedTransaction(_, _)
+                    | ClientError::OversizedTransactionBundleBytes(_, _)
+                    | ClientError::OversizedTransactionBundleCount(_, _) => {
+                        if cfg!(debug_assertions) {
+                            panic!("{}", msg);
+                        } else {
+                            error!("{}", msg);
```

Comment:
Better to inline the vars when possible https://rust-lang.github.io/rust-clippy/master/index.html#uninlined_format_args, there are a few more other places